core/src/main/scala/cats/data/Catenable.scala

```diff
@@ -0,0 +1,473 @@
+        case Empty =>
+          if (rights.isEmpty) {
+            result = None
+          } else {
+            c = rights.last
+            rights.trimEnd(1)
+          }
+      }
+    }
+    // scalastyle:on null
+    result
+  }
+
+  /** Returns true if there are no elements in this collection. */
+  def isEmpty: Boolean
+
+  /** Returns false if there are no elements in this collection. */
+  def nonEmpty: Boolean = !isEmpty
+
+  /** Concatenates this with `c` in O(1) runtime. */
+  final def ++[A2 >: A](c: Catenable[A2]): Catenable[A2] =
+    append(this, c)
+
+  /** Returns a new catenable consisting of `a` followed by this. O(1) runtime. */
+  final def cons[A2 >: A](a: A2): Catenable[A2] =
+    append(singleton(a), this)
+
+  /** Alias for [[cons]]. */
+  final def +:[A2 >: A](a: A2): Catenable[A2] =
+    cons(a)
+
+  /** Returns a new catenable consisting of this followed by `a`. O(1) runtime. */
+  final def snoc[A2 >: A](a: A2): Catenable[A2] =
+    append(this, singleton(a))
+
+  /** Alias for [[snoc]]. */
+  final def :+[A2 >: A](a: A2): Catenable[A2] =
+    snoc(a)
+
+  /** Applies the supplied function to each element and returns a new catenable. */
+  final def map[B](f: A => B): Catenable[B] =
+    fromSeq(iterator.map(f).toVector)
+
+  /** Applies the supplied function to each element and returns a new catenable from the concatenated results */
+  final def flatMap[B](f: A => Catenable[B]): Catenable[B] =
+    foldLeft(nil: Catenable[B])((acc, a) => acc ++ f(a))
+
+  /** Folds over the elements from left to right using the supplied initial value and function. */
+  final def foldLeft[B](z: B)(f: (B, A) => B): B = {
+    var result = z
+    foreach(a => result = f(result, a))
+    result
+  }
+
+  /** Folds over the elements from right to left using the supplied initial value and function. */
+  final def foldRight[B](z: B)(f: (A, B) => B): B = {
+    val stack = new ArrayStack[A]
+    foreach { a => stack += a; () }
+    var result = z
+    while (!stack.isEmpty) { result = f(stack.pop, result) }
+    result
+  }
+
+  /** Collect `B` from this for which `f` is defined */
+  final def collect[B](pf: PartialFunction[A, B]): Catenable[B] =
+    foldLeft(Catenable.nil: Catenable[B]) { (acc, a) =>
+      // trick from TraversableOnce, used to avoid calling both isDefined and apply (or calling lift)
+      val x = pf.applyOrElse(a, sentinel)
+      if (x.asInstanceOf[AnyRef] ne sentinel) acc :+ x.asInstanceOf[B]
+      else acc
+    }
+
+  /** Remove elements not matching the predicate */
+  final def filter(f: A => Boolean): Catenable[A] =
+    collect { case a if f(a) => a }
+
+  /** Remove elements matching the predicate */
+  final def filterNot(f: A => Boolean): Catenable[A] =
+    filter(a => !f(a))
+
+  /** Find the first element matching the predicate, if one exists */
+  final def find(f: A => Boolean): Option[A] = {
+    var result: Option[A] = Option.empty[A]
+    foreachUntil { a =>
```

Comment:
Can do with iterator + while

Reply:
The foreach version is faster than this one for some reason

```scala
final def findIterator(f: A => Boolean): Option[A] = {
    val iter = iterator
    var result: Option[A] = Option.empty[A]
    while (iter.hasNext && result.isEmpty) {
      val a = iter.next
      if (f(a)) result = Some(a)
    }
    result
  }
```

```
[info] Benchmark                 Mode  Cnt       Score      Error  Units
[info] ChainBench.find          thrpt   15  141200.957 ± 2605.379  ops/s
[info] ChainBench.findIterator  thrpt   15   62271.920 ±  666.115  ops/s
```